examples/ilu-preconditioned-solver-iterative-trisolves/ilu-preconditioned-solver-iterative-trisolves.cpp

```diff
@@ -0,0 +1,146 @@
+               gko::HipExecutor::get_num_devices() > 0) {
+        exec = gko::HipExecutor::create(0, gko::OmpExecutor::create());
+    } else {
+        std::cerr << "Usage: " << argv[0] << " [executor]" << std::endl;
+        std::exit(-1);
+    }
+
+    // Read data
+    auto A = gko::share(gko::read<mtx>(std::ifstream("data/A.mtx"), exec));
+    auto b = gko::read<vec>(std::ifstream("data/b.mtx"), exec);
+    auto x = gko::read<vec>(std::ifstream("data/x0.mtx"), exec);
+
+    // Generate incomplete factors using ParILU
+    auto par_ilu_fact =
+        gko::factorization::ParIlu<ValueType, IndexType>::build().on(exec);
+    // Generate concrete factorization for input matrix
+    auto par_ilu = par_ilu_fact->generate(A);
+
+    // Generate an iterative refinement factory to be used as triangular solver
+    // in the preconditioner application. The generated method is equivalent
+    // to doing five block-Jacobi sweeps with maximum block size 16.
+    auto trisolve_factory =
+        ir::build()
+            .with_solver(bj::build()
+                             .with_max_block_size(16u)
+                             .with_storage_optimization(
+                                 gko::precision_reduction::autodetect())
+                             .on(exec))
+            .with_criteria(
+                gko::stop::Iteration::build().with_max_iters(5u).on(exec))
+            .on(exec);
+
+    // Generate an ILU preconditioner factory by setting lower and upper
+    // triangular solver - in this case the previously defined iterative
+    // refinement method.
+    auto ilu_pre_factory =
+        gko::preconditioner::Ilu<ir, ir, false>::build()
+            .with_l_solver_factory(gko::clone(trisolve_factory))
+            .with_u_solver_factory(gko::clone(trisolve_factory))
+            .on(exec);
+
+    // Use incomplete factors to generate ILU preconditioner
+    auto ilu_preconditioner = ilu_pre_factory->generate(gko::share(par_ilu));
+
+    // Use preconditioner inside GMRES solver factory
+    // Generating a solver factory tied to a specific preconditioner makes sense
+    // if there are several very similar systems to solve, and the same
+    // solver+preconditioner combination is expected to be effective.
+    const gko::remove_complex<ValueType> reduction_factor = 1e-12;
+    auto ilu_gmres_factory =
+        gmres::build()
+            .with_criteria(
+                gko::stop::Iteration::build().with_max_iters(1000u).on(exec),
+                gko::stop::ResidualNormReduction<ValueType>::build()
+                    .with_reduction_factor(reduction_factor)
+                    .on(exec))
+            .with_generated_preconditioner(gko::share(ilu_preconditioner))
+            .on(exec);
+
+    // Generate preconditioned solver for a specific target system
+    auto ilu_gmres = ilu_gmres_factory->generate(A);
+
+    // Solve system
+    ilu_gmres->apply(gko::lend(b), gko::lend(x));
```

Comment:
```suggestion
    ilu_gmres->apply(lend(b), lend(x));
```
That's thanks to the magic of Argument-Dependent Lookup (ADL). Same for all `gko::share`.
This works because `lend` is not defined in the current scope, but it is defined in the namespace that defines a template parameter of the type of `b` (namely `std::unique_ptr<gko::Dense<...>>`)